qiskit/transpiler/passes/layout/sabre_layout.py

```diff
@@ -0,0 +1,147 @@
+    for NISQ-era quantum devices." ASPLOS 2019.
+    `arXiv:1809.02573 <https://arxiv.org/pdf/1809.02573.pdf>`_
+    """
+
+    def __init__(self, coupling_map, routing_pass=None, seed=None,
+                 max_iterations=3):
+        """SabreLayout initializer.
+
+        Args:
+            coupling_map (Coupling): directed graph representing a coupling map.
+            routing_pass (BasePass): the routing pass to use while iterating.
+            seed (int): seed for setting a random first trial layout.
+            max_iterations (int): number of forward-backward iterations.
+        """
+        super().__init__()
+        self.coupling_map = coupling_map
+        self.routing_pass = routing_pass
+        self.seed = seed
+        self.max_iterations = max_iterations
+
+    def run(self, dag):
+        """Run the SabreLayout pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to find layout for.
+
+        Raises:
+            TranspilerError: if dag wider than self.coupling_map
+        """
+        def _layout_and_route_passmanager(initial_layout):
+            """Return a passmanager for a full layout and routing.
+
+            We use a factory to remove potential statefulness of passes.
+            """
+            layout_and_route = [SetLayout(initial_layout),
+                                FullAncillaAllocation(self.coupling_map),
+                                EnlargeWithAncilla(),
+                                ApplyLayout(),
+                                self.routing_pass]
+            pm = PassManager(layout_and_route)
+            return pm
+
+        def _compose_layouts(initial_layout, pass_final_layout, qregs):
+            """Return the real final_layout resulting from the composition
+            of an initial_layout with the final_layout reported by a pass.
+
+            The routing passes internally start with a trivial layout, as the
+            layout gets applied to the circuit prior to running them. So the
+            "final_layout" they report must be amended to account for the actual
+            initial_layout that was selected.
+            """
+            trivial_layout = Layout.generate_trivial_layout(*qregs)
+            pass_final_layout = Layout({trivial_layout[v.index]: p
+                                        for v, p in pass_final_layout.get_virtual_bits().items()})
+            qubit_map = Layout.combine_into_edge_map(initial_layout, trivial_layout)
+            final_layout = {v: pass_final_layout[qubit_map[v]]
+                            for v, _ in initial_layout.get_virtual_bits().items()}
+            return Layout(final_layout)
+
+        if len(dag.qubits) > self.coupling_map.size():
+            raise TranspilerError('More virtual qubits exist than physical.')
+
+        # Choose a random initial_layout.
+        if self.seed is None:
+            self.seed = np.random.randint(0, np.iinfo(np.int32).max)
+        rng = np.random.default_rng(self.seed)
```

Comment:
```suggestion
        rng = np.random.RandomState(self.seed)
```

Reply:
no this needs to be like this i keep saying this!
https://github.com/Qiskit/qiskit-terra/pull/3586#discussion_r421211627

Reply:
I keep missing them up. Sorry.
```suggestion
        rng = np.random.default_rng(self.seed)
```